File: mobile-sdk/flutter-sdk/lib/pyairtable_client.dart
import 'dart:async';
import 'package:rxdart/rxdart.dart';

import 'auth/auth_service.dart';
import 'api/api_client.dart';
import 'websocket/websocket_client.dart';
import 'offline/offline_manager.dart';
import 'notifications/notification_manager.dart';
import 'types/types.dart';
import 'utils/constants.dart';

class PyAirtableClient {
  final PyAirtableConfig config;
  late final AuthService _authService;
  late final ApiClient _apiClient;
  WebSocketClient? _webSocketClient;
  OfflineManager? _offlineManager;
  NotificationManager? _notificationManager;
  
  bool _isInitialized = false;
  
  // Event streams
  final _loginController = StreamController<User>.broadcast();
  final _logoutController = StreamController<void>.broadcast();
  final _onlineController = StreamController<bool>.broadcast();
  final _syncController = StreamController<SyncEvent>.broadcast();
  final _recordController = StreamController<RecordEvent>.broadcast();
  final _webSocketController = StreamController<WebSocketEvent>.broadcast();

  // Public streams
  Stream<User> get onLogin => _loginController.stream;
  Stream<void> get onLogout => _logoutController.stream;
  Stream<bool> get onNetworkChange => _onlineController.stream;
  Stream<SyncEvent> get onSyncEvent => _syncController.stream;
  Stream<RecordEvent> get onRecordEvent => _recordController.stream;
  Stream<WebSocketEvent> get onWebSocketEvent => _webSocketController.stream;

  PyAirtableClient({required this.config}) {
    if (config.apiKey.isEmpty || config.baseUrl.isEmpty) {
      throw const ValidationException('API key and base URL are required');
    }

    _authService = AuthService(
      baseUrl: config.baseUrl,
      apiKey: config.apiKey,
    );
    
    _apiClient = ApiClient(
      authService: _authService,
      config: config,
    );
    
    // Initialize optional services
    if (config.enableWebSocket) {
      _webSocketClient = WebSocketClient(
        authService: _authService,
        baseUrl: config.baseUrl,
      );
      _setupWebSocketListeners();
    }
    
    if (config.enableOfflineSync) {
      _offlineManager = OfflineManager(
        apiClient: _apiClient,
        config: const OfflineManagerConfig(),
      );
      _setupOfflineListeners();
    }
    
    if (config.enablePushNotifications) {
      _notificationManager = NotificationManager();
    }
  }

  /// Initialize the client
  Future<void> initialize() async {
    if (_isInitialized) return;

    try {
      // Initialize auth service
      await _authService.initialize();
      
      // Initialize notification manager if enabled
      if (_notificationManager != null) {
        await _notificationManager!.initialize();
      }
      
      // Initialize offline manager if enabled
      if (_offlineManager != null) {
        await _offlineManager!.initialize();
      }

      _isInitialized = true;
    } catch (error) {
      throw PyAirtableException('Failed to initialize PyAirtable client: $error');
    }
  }

  // Authentication Methods

  /// Login with credentials
  Future<User> login(LoginCredentials credentials) async {
    await _authService.login(credentials);
    final user = _authService.user;
    
    if (user == null) {
      throw const PyAirtableException('Login succeeded but no user data received');
    }

    // Connect WebSocket if enabled
    if (_webSocketClient != null) {
      try {
        await _webSocketClient!.connect();
      } catch (error) {
        print('WebSocket connection failed: $error');
      }
    }

    _loginController.add(user);
    return user;
  }

  /// Logout
  Future<void> logout() async {
    // Disconnect WebSocket
    _webSocketClient?.disconnect();

    // Clear offline data
    if (_offlineManager != null) {
      await _offlineManager!.clearCache();
    }

    await _authService.logout();
    _logoutController.add(null);
  }

  /// Get current user
  User? get currentUser => _authService.user;

  /// Check if authenticated
  bool get isAuthenticated => _authService.isAuthenticated;

  // Base Operations

  /// Get all accessible bases
  Future<List<Base>> getBases() async {
    return _apiClient.getBases();
  }

  /// Get a specific base
  Future<Base> getBase(String baseId) async {
    return _apiClient.getBase(baseId);
  }

  // Table Operations

  /// Get all tables in a base
  Future<List<Table>> getTables(String baseId) async {
    return _apiClient.getTables(baseId);
  }

  /// Get a specific table
  Future<Table> getTable(String baseId, String tableId) async {
    return _apiClient.getTable(baseId, tableId);
  }

  // Record Operations

  /// List records with optional caching
  Future<RecordsResult> listRecords(
    String baseId,
    String tableId, {
    RecordQuery? query,
    bool useCache = true,
  }) async {
    // Try to get from cache first if offline manager is enabled
    if (useCache && _offlineManager != null && (!_apiClient.isOnline || config.enableOfflineSync)) {
      final cachedRecords = await _offlineManager!.getCachedRecords(baseId, tableId);
      if (cachedRecords.isNotEmpty || !_apiClient.isOnline) {
        return RecordsResult(records: cachedRecords);
      }
    }

    // Fetch from API
    final result = await _apiClient.listRecords(baseId, tableId, query: query);
    
    // Cache the results
    if (_offlineManager != null) {
      await _offlineManager!.cacheRecords(baseId, tableId, result.records);
    }

    return result;
  }

  /// Get a specific record
  Future<Record> getRecord(String baseId, String tableId, String recordId) async {
    return _apiClient.getRecord(baseId, tableId, recordId);
  }

  /// Create a new record (with offline support)
  Future<Record> createRecord(
    String baseId,
    String tableId,
    RecordCreate record,
  ) async {
    if (_offlineManager != null) {
      return _offlineManager!.createRecord(baseId, tableId, record);
    }
    
    return _apiClient.createRecord(baseId, tableId, record);
  }

  /// Create multiple records
  Future<List<Record>> createRecords(
    String baseId,
    String tableId,
    List<RecordCreate> records,
  ) async {
    return _apiClient.createRecords(baseId, tableId, records);
  }

  /// Update a record (with offline support)
  Future<Record> updateRecord(
    String baseId,
    String tableId,
    RecordUpdate record,
  ) async {
    if (_offlineManager != null) {
      return _offlineManager!.updateRecord(baseId, tableId, record);
    }
    
    return _apiClient.updateRecord(baseId, tableId, record);
  }

  /// Update multiple records
  Future<List<Record>> updateRecords(
    String baseId,
    String tableId,
    List<RecordUpdate> records,
  ) async {
    return _apiClient.updateRecords(baseId, tableId, records);
  }

  /// Delete a record (with offline support)
  Future<void> deleteRecord(String baseId, String tableId, String recordId) async {
    if (_offlineManager != null) {
      return _offlineManager!.deleteRecord(baseId, tableId, recordId);
    }
    
    return _apiClient.deleteRecord(baseId, tableId, recordId);
  }

  /// Delete multiple records
  Future<void> deleteRecords(String baseId, String tableId, List<String> recordIds) async {
    return _apiClient.deleteRecords(baseId, tableId, recordIds);
  }

  // Formula Operations

  /// Evaluate a formula
  Future<FormulaResult> evaluateFormula(
    String baseId,
    String tableId,
    String recordId,
    String formula,
  ) async {
    try {
      final result = await _apiClient.evaluateFormula(baseId, tableId, recordId, formula);
      return FormulaResult(value: result);
    } catch (error) {
      return FormulaResult(
        value: null,
        error: error is Exception ? error.toString() : 'Formula evaluation failed',
      );
    }
  }

  // File Operations

  /// Upload a file with progress tracking
  Future<FileUploadResult> uploadFile(
    String filePath,
    String filename, {
    void Function(FileUploadProgress)? onProgress,
  }) async {
    return _apiClient.uploadFile(filePath, filename, onProgress: onProgress);
  }

  /// Download a file
  Future<List<int>> downloadFile(String fileId) async {
    return _apiClient.downloadFile(fileId);
  }

  // WebSocket Operations

  /// Subscribe to real-time updates for a table
  void subscribeToTable(String baseId, String tableId) {
    if (_webSocketClient == null) {
      throw const PyAirtableException('WebSocket not enabled');
    }
    
    _webSocketClient!.subscribeToTable(baseId, tableId);
  }

  /// Unsubscribe from real-time updates for a table
  void unsubscribeFromTable(String baseId, String tableId) {
    if (_webSocketClient == null) {
      throw const PyAirtableException('WebSocket not enabled');
    }
    
    _webSocketClient!.unsubscribeFromTable(baseId, tableId);
  }

  /// Check if WebSocket is connected
  bool get isWebSocketConnected => _webSocketClient?.isConnected ?? false;

  // Offline Operations

  /// Force sync with server
  Future<void> sync() async {
    if (_offlineManager == null) {
      throw const PyAirtableException('Offline sync not enabled');
    }
    
    return _offlineManager!.sync();
  }

  /// Get pending operations count
  int get pendingOperationsCount => _offlineManager?.pendingOperationsCount ?? 0;

  /// Check if device is online
  bool get isOnline => _apiClient.isOnline;

  // Notification Operations

  /// Show a notification
  Future<void> showNotification(String title, String body, {Map<String, dynamic>? data}) async {
    if (_notificationManager == null) {
      throw const PyAirtableException('Push notifications not enabled');
    }
    
    return _notificationManager!.showNotification(title, body, data: data);
  }

  // Event Setup

  void _setupWebSocketListeners() {
    if (_webSocketClient == null) return;

    _webSocketClient!.onConnect.listen((_) {
      _webSocketController.add(const WebSocketEvent('connect', {}));
    });

    _webSocketClient!.onDisconnect.listen((_) {
      _webSocketController.add(const WebSocketEvent('disconnect', {}));
    });

    _webSocketClient!.onError.listen((error) {
      _webSocketController.add(WebSocketEvent('error', {'error': error.toString()}));
    });

    _webSocketClient!.onMessage.listen((message) {
      switch (message.type) {
        case WebSocketEvents.recordCreated:
          _recordController.add(RecordEvent('created', message.data));
          if (_notificationManager != null) {
            _showRecordNotification('created', message.data['tableName'] ?? 'table');
          }
          break;
        case WebSocketEvents.recordUpdated:
          _recordController.add(RecordEvent('updated', message.data));
          if (_notificationManager != null) {
            _showRecordNotification('updated', message.data['tableName'] ?? 'table');
          }
          break;
        case WebSocketEvents.recordDeleted:
          _recordController.add(RecordEvent('deleted', message.data));
          if (_notificationManager != null) {
            _showRecordNotification('deleted', message.data['tableName'] ?? 'table');
          }
          break;
      }
    });
  }

  void _setupOfflineListeners() {
    if (_offlineManager == null) return;

    _offlineManager!.onNetworkChange.listen((isOnline) {
      _onlineController.add(isOnline);
    });

    _offlineManager!.onSyncStart.listen((_) {
      _syncController.add(const SyncEvent('start', {}));
      if (_notificationManager != null) {
        _showSyncNotification('syncing');
      }
    });

    _offlineManager!.onSyncComplete.listen((_) {
      _syncController.add(const SyncEvent('complete', {}));
      if (_notificationManager != null) {
        _showSyncNotification('completed');
      }
    });

    _offlineManager!.onSyncError.listen((error) {
      _syncController.add(SyncEvent('error', {'error': error.toString()}));
      if (_notificationManager != null) {
        _showSyncNotification('failed', error.toString());
      }
    });
  }

  void _showRecordNotification(String action, String tableName) {
    _notificationManager?.showNotification(
      'Record ${action.toUpperCase()}',
      'A record was $action in $tableName',
      data: {'type': 'record_update', 'action': action, 'tableName': tableName},
    );
  }

  void _showSyncNotification(String status, [String? details]) {
    final messages = {
      'syncing': 'Syncing data with server...',
      'completed': details ?? 'Data synchronized successfully',
      'failed': details ?? 'Sync failed. Will retry automatically.',
    };

    _notificationManager?.showNotification(
      'PyAirtable Sync',
      messages[status] ?? 'Sync status update',
      data: {'type': 'sync', 'status': status},
    );
  }

  /// Cleanup and destroy all resources
  Future<void> dispose() async {
    _webSocketClient?.disconnect();
    await _offlineManager?.dispose();
    await _notificationManager?.dispose();
    
    await _loginController.close();
    await _logoutController.close();
    await _onlineController.close();
    await _syncController.close();
    await _recordController.close();
    await _webSocketController.close();
    
    _isInitialized = false;
  }
}

// Helper classes for events
class SyncEvent {
  final String type;
  final Map<String, dynamic> data;

  const SyncEvent(this.type, this.data);
}

class RecordEvent {
  final String type;
  final Map<String, dynamic> data;

  const RecordEvent(this.type, this.data);
}

class WebSocketEvent {
  final String type;
  final Map<String, dynamic> data;

  const WebSocketEvent(this.type, this.data);
}

class RecordsResult {
  final List<Record> records;
  final String? offset;

  const RecordsResult({required this.records, this.offset});
}

class FormulaResult {
  final dynamic value;
  final String? error;

  const FormulaResult({this.value, this.error});
}

class FileUploadResult {
  final String id;
  final String url;

  const FileUploadResult({required this.id, required this.url});
}

File: mobile-sdk/react-native-sdk/src/PyAirtableClient.ts
import { EventEmitter } from 'events';
import { AuthService } from './auth/AuthService';
import { ApiClient } from './api/ApiClient';
import { WebSocketClient } from './websocket/WebSocketClient';
import { OfflineManager } from './offline/OfflineManager';
import { NotificationManager } from './notifications/NotificationManager';
import { 
  PyAirtableConfig,
  LoginCredentials,
  User,
  Base,
  Record,
  RecordCreate,
  RecordUpdate,
  RecordQuery,
  Table,
  FormulaResult,
  FileUploadProgress
} from './types';
import { DEFAULT_CONFIG, WEBSOCKET_EVENTS } from './utils/constants';
import { PyAirtableSDKError, ValidationError } from './utils/errors';

export class PyAirtableClient extends EventEmitter {
  private config: Required<PyAirtableConfig>;
  private authService: AuthService;
  private apiClient: ApiClient;
  private webSocketClient?: WebSocketClient;
  private offlineManager?: OfflineManager;
  private notificationManager?: NotificationManager;
  private isInitialized = false;

  constructor(config: PyAirtableConfig) {
    super();
    
    if (!config.apiKey || !config.baseUrl) {
      throw new ValidationError('API key and base URL are required');
    }

    this.config = { ...DEFAULT_CONFIG, ...config };
    
    // Initialize core services
    this.authService = new AuthService(this.config.baseUrl, this.config.apiKey);
    this.apiClient = new ApiClient(this.authService, this.config);
    
    // Initialize optional services
    if (this.config.enableWebSocket) {
      this.webSocketClient = new WebSocketClient(this.authService, {
        baseUrl: this.config.baseUrl,
      });
      this.setupWebSocketListeners();
    }
    
    if (this.config.enableOfflineSync) {
      this.offlineManager = new OfflineManager(this.apiClient);
      this.setupOfflineListeners();
    }
    
    if (this.config.enablePushNotifications) {
      this.notificationManager = new NotificationManager();
    }
  }

  /**
   * Initialize the client
   */
  async initialize(): Promise<void> {
    if (this.isInitialized) return;

    try {
      // Initialize notification manager if enabled
      if (this.notificationManager) {
        await this.notificationManager.initialize();
      }

      this.isInitialized = true;
      this.emit('initialized');
    } catch (error) {
      throw new PyAirtableSDKError('Failed to initialize PyAirtable client: ' + error.message);
    }
  }

  // Authentication Methods

  /**
   * Login with credentials
   */
  async login(credentials: LoginCredentials): Promise<User> {
    await this.authService.login(credentials);
    const user = this.authService.getUser();
    
    if (!user) {
      throw new PyAirtableSDKError('Login succeeded but no user data received');
    }

    // Connect WebSocket if enabled
    if (this.webSocketClient) {
      try {
        await this.webSocketClient.connect();
      } catch (error) {
        console.warn('WebSocket connection failed:', error);
      }
    }

    this.emit('login', user);
    return user;
  }

  /**
   * Logout
   */
  async logout(): Promise<void> {
    // Disconnect WebSocket
    if (this.webSocketClient) {
      this.webSocketClient.disconnect();
    }

    // Clear offline data
    if (this.offlineManager) {
      await this.offlineManager.clearCache();
    }

    await this.authService.logout();
    this.emit('logout');
  }

  /**
   * Get current user
   */
  getCurrentUser(): User | null {
    return this.authService.getUser();
  }

  /**
   * Check if authenticated
   */
  isAuthenticated(): boolean {
    return this.authService.isAuthenticated();
  }

  // Base Operations

  /**
   * Get all accessible bases
   */
  async getBases(): Promise<Base[]> {
    return this.apiClient.getBases();
  }

  /**
   * Get a specific base
   */
  async getBase(baseId: string): Promise<Base> {
    return this.apiClient.getBase(baseId);
  }

  // Table Operations

  /**
   * Get all tables in a base
   */
  async getTables(baseId: string): Promise<Table[]> {
    return this.apiClient.getTables(baseId);
  }

  /**
   * Get a specific table
   */
  async getTable(baseId: string, tableId: string): Promise<Table> {
    return this.apiClient.getTable(baseId, tableId);
  }

  // Record Operations

  /**
   * List records with optional caching
   */
  async listRecords(
    baseId: string,
    tableId: string,
    query?: RecordQuery,
    useCache: boolean = true
  ): Promise<{ records: Record[]; offset?: string }> {
    // Try to get from cache first if offline or cache enabled
    if (useCache && this.offlineManager && (!this.apiClient.isDeviceOnline() || this.config.enableOfflineSync)) {
      const cachedRecords = await this.offlineManager.getCachedRecords(baseId, tableId);
      if (cachedRecords.length > 0 || !this.apiClient.isDeviceOnline()) {
        return { records: cachedRecords };
      }
    }

    // Fetch from API
    const result = await this.apiClient.listRecords(baseId, tableId, query);
    
    // Cache the results
    if (this.offlineManager) {
      await this.offlineManager.cacheRecords(baseId, tableId, result.records);
    }

    return result;
  }

  /**
   * Get a specific record
   */
  async getRecord(baseId: string, tableId: string, recordId: string): Promise<Record> {
    return this.apiClient.getRecord(baseId, tableId, recordId);
  }

  /**
   * Create a new record (with offline support)
   */
  async createRecord(
    baseId: string,
    tableId: string,
    record: RecordCreate
  ): Promise<Record> {
    if (this.offlineManager) {
      return this.offlineManager.createRecord(baseId, tableId, record);
    }
    
    return this.apiClient.createRecord(baseId, tableId, record);
  }

  /**
   * Create multiple records
   */
  async createRecords(
    baseId: string,
    tableId: string,
    records: RecordCreate[]
  ): Promise<Record[]> {
    // For bulk operations, use API directly (offline manager handles individual records)
    return this.apiClient.createRecords(baseId, tableId, records);
  }

  /**
   * Update a record (with offline support)
   */
  async updateRecord(
    baseId: string,
    tableId: string,
    record: RecordUpdate
  ): Promise<Record> {
    if (this.offlineManager) {
      return this.offlineManager.updateRecord(baseId, tableId, record);
    }
    
    return this.apiClient.updateRecord(baseId, tableId, record);
  }

  /**
   * Update multiple records
   */
  async updateRecords(
    baseId: string,
    tableId: string,
    records: RecordUpdate[]
  ): Promise<Record[]> {
    return this.apiClient.updateRecords(baseId, tableId, records);
  }

  /**
   * Delete a record (with offline support)
   */
  async deleteRecord(baseId: string, tableId: string, recordId: string): Promise<void> {
    if (this.offlineManager) {
      return this.offlineManager.deleteRecord(baseId, tableId, recordId);
    }
    
    return this.apiClient.deleteRecord(baseId, tableId, recordId);
  }

  /**
   * Delete multiple records
   */
  async deleteRecords(baseId: string, tableId: string, recordIds: string[]): Promise<void> {
    return this.apiClient.deleteRecords(baseId, tableId, recordIds);
  }

  // Formula Operations

  /**
   * Evaluate a formula
   */
  async evaluateFormula(
    baseId: string,
    tableId: string,
    recordId: string,
    formula: string
  ): Promise<FormulaResult> {
    try {
      const result = await this.apiClient.evaluateFormula(baseId, tableId, recordId, formula);
      return { value: result };
    } catch (error) {
      return { 
        value: null, 
        error: error instanceof Error ? error.message : 'Formula evaluation failed' 
      };
    }
  }

  // File Operations

  /**
   * Upload a file with progress tracking
   */
  async uploadFile(
    file: File | Blob,
    filename: string,
    onProgress?: (progress: FileUploadProgress) => void
  ): Promise<{ id: string; url: string }> {
    return this.apiClient.uploadFile(file, filename, (progress) => {
      if (onProgress) {
        onProgress({
          loaded: progress.loaded,
          total: progress.total,
          percentage: Math.round((progress.loaded / progress.total) * 100),
        });
      }
    });
  }

  /**
   * Download a file
   */
  async downloadFile(fileId: string): Promise<Blob> {
    return this.apiClient.downloadFile(fileId);
  }

  // WebSocket Operations

  /**
   * Subscribe to real-time updates for a table
   */
  subscribeToTable(baseId: string, tableId: string): void {
    if (!this.webSocketClient) {
      throw new PyAirtableSDKError('WebSocket not enabled');
    }
    
    this.webSocketClient.subscribeToTable(baseId, tableId);
  }

  /**
   * Unsubscribe from real-time updates for a table
   */
  unsubscribeFromTable(baseId: string, tableId: string): void {
    if (!this.webSocketClient) {
      throw new PyAirtableSDKError('WebSocket not enabled');
    }
    
    this.webSocketClient.unsubscribeFromTable(baseId, tableId);
  }

  /**
   * Check if WebSocket is connected
   */
  isWebSocketConnected(): boolean {
    return this.webSocketClient?.isConnected() ?? false;
  }

  // Offline Operations

  /**
   * Force sync with server
   */
  async sync(): Promise<void> {
    if (!this.offlineManager) {
      throw new PyAirtableSDKError('Offline sync not enabled');
    }
    
    return this.offlineManager.sync();
  }

  /**
   * Get pending operations count
   */
  getPendingOperationsCount(): number {
    return this.offlineManager?.getPendingOperationsCount() ?? 0;
  }

  /**
   * Check if device is online
   */
  isOnline(): boolean {
    return this.apiClient.isDeviceOnline();
  }

  // Notification Operations

  /**
   * Show a notification
   */
  showNotification(title: string, body: string, data?: any): void {
    if (!this.notificationManager) {
      throw new PyAirtableSDKError('Push notifications not enabled');
    }
    
    this.notificationManager.showNotification({
      title,
      body,
      data,
    });
  }

  /**
   * Set badge number (iOS)
   */
  setBadgeNumber(number: number): void {
    if (!this.notificationManager) {
      throw new PyAirtableSDKError('Push notifications not enabled');
    }
    
    this.notificationManager.setBadgeNumber(number);
  }

  // Event Setup

  /**
   * Setup WebSocket event listeners
   */
  private setupWebSocketListeners(): void {
    if (!this.webSocketClient) return;

    this.webSocketClient.on(WEBSOCKET_EVENTS.CONNECT, () => {
      this.emit('websocket:connect');
    });

    this.webSocketClient.on(WEBSOCKET_EVENTS.DISCONNECT, () => {
      this.emit('websocket:disconnect');
    });

    this.webSocketClient.on(WEBSOCKET_EVENTS.ERROR, (error) => {
      this.emit('websocket:error', error);
    });

    this.webSocketClient.on(WEBSOCKET_EVENTS.RECORD_CREATED, (data) => {
      this.emit('record:created', data);
      
      if (this.notificationManager) {
        this.notificationManager.showRecordUpdateNotification('created', data.tableName);
      }
    });

    this.webSocketClient.on(WEBSOCKET_EVENTS.RECORD_UPDATED, (data) => {
      this.emit('record:updated', data);
      
      if (this.notificationManager) {
        this.notificationManager.showRecordUpdateNotification('updated', data.tableName);
      }
    });

    this.webSocketClient.on(WEBSOCKET_EVENTS.RECORD_DELETED, (data) => {
      this.emit('record:deleted', data);
      
      if (this.notificationManager) {
        this.notificationManager.showRecordUpdateNotification('deleted', data.tableName);
      }
    });
  }

  /**
   * Setup offline manager event listeners
   */
  private setupOfflineListeners(): void {
    if (!this.offlineManager) return;

    this.offlineManager.on('online', () => {
      this.emit('online');
    });

    this.offlineManager.on('offline', () => {
      this.emit('offline');
    });

    this.offlineManager.on('syncStart', () => {
      this.emit('sync:start');
      
      if (this.notificationManager) {
        this.notificationManager.showSyncNotification('syncing');
      }
    });

    this.offlineManager.on('syncComplete', () => {
      this.emit('sync:complete');
      
      if (this.notificationManager) {
        this.notificationManager.showSyncNotification('completed');
      }
    });

    this.offlineManager.on('syncError', (error) => {
      this.emit('sync:error', error);
      
      if (this.notificationManager) {
        this.notificationManager.showSyncNotification('failed', error.message);
      }
    });
  }

  /**
   * Cleanup and destroy all resources
   */
  destroy(): void {
    this.webSocketClient?.destroy();
    this.offlineManager?.destroy();
    this.notificationManager?.destroy();
    
    this.removeAllListeners();
    this.isInitialized = false;
  }
}